src/ec/curve25519/ops.rs

```diff
@@ -0,0 +1,120 @@
+// Copyright 2017 Brian Smith.
+//
+// Permission to use, copy, modify, and/or distribute this software for any
+// purpose with or without fee is hereby granted, provided that the above
+// copyright notice and this permission notice appear in all copies.
+//
+// THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHORS DISCLAIM ALL WARRANTIES
+// WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+// MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY
+// SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+// WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION
+// OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
+// CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+
+//! Elliptic curve operations on Curve25519.
+
+use {bssl, c, error};
+
+// Keep this in sync with `fe` in curve25519/internal.h.
+pub type Elem = [i32; ELEM_LIMBS];
+const ELEM_LIMBS: usize = 10;
+
+type EncodedPoint = [u8; ELEM_LEN];
+pub const ELEM_LEN: usize = 32;
+
+pub type Scalar = [u8; SCALAR_LEN];
+pub const SCALAR_LEN: usize = 32;
+
+pub type UnreducedScalar = [u8; UNREDUCED_SCALAR_LEN];
+const UNREDUCED_SCALAR_LEN: usize = SCALAR_LEN * 2;
+
+// Keep this in sync with `ge_p3` in curve25519/internal.h.
+#[repr(C)]
+pub struct ExtPoint {
+    x: Elem,
+    y: Elem,
+    z: Elem,
+    t: Elem,
+}
+
+impl ExtPoint {
+    pub fn new_at_infinity() -> Self {
+        ExtPoint {
+            x: [0; ELEM_LIMBS],
+            y: [0; ELEM_LIMBS],
+            z: [0; ELEM_LIMBS],
+            t: [0; ELEM_LIMBS],
+        }
+    }
+
+    pub fn decode_vartime(encoded: &EncodedPoint)
+                          -> Result<Self, error::Unspecified> {
+        let mut point = Self::new_at_infinity();
+
+        try!(bssl::map_result(unsafe {
+            GFp_x25519_ge_frombytes_vartime(&mut point, encoded)
+        }));
+
+        Ok(point)
+    }
+
+    pub fn encode(&self) -> EncodedPoint {
+        encode_point(&self.x, &self.y, &self.z)
+    }
+
+    pub fn invert_vartime(&mut self) {
+        for i in 0..ELEM_LIMBS {
+            self.x[i] = -self.x[i];
+            self.t[i] = -self.t[i];
+        }
+    }
+}
+
+// Keep this in sync with `ge_p2` in curve25519/internal.h.
+#[repr(C)]
+pub struct Point {
+    x: Elem,
+    y: Elem,
+    z: Elem,
+}
+
+impl Point {
+    pub fn new_at_infinity() -> Self {
+        Point {
+            x: [0; ELEM_LIMBS],
+            y: [0; ELEM_LIMBS],
+            z: [0; ELEM_LIMBS],
+        }
+    }
+
+    pub fn encode(&self) -> EncodedPoint {
+        encode_point(&self.x, &self.y, &self.z)
+    }
+}
+
+fn encode_point(x: &Elem, y: &Elem, z: &Elem) -> [u8; ELEM_LEN] {
```

Comment:
As I noted in another comment, I think we should give some descriptive type alias to this `[u8; ELEM_LEN]`.